File: bindings/java/src/main/java/com/deftio/triepack/Values.java
/*
 * Values.java
 *
 * Value encoder/decoder for 8 type tags -- matches src/core/value.c.
 *
 * Copyright (c) 2026 M. A. Chatterjee
 * SPDX-License-Identifier: BSD-2-Clause
 */

package com.deftio.triepack;

import java.nio.charset.StandardCharsets;

/**
 * Encode and decode typed values to/from a bitstream.
 */
public class Values {

    // Type tags (4-bit, matches tp_value_type enum)
    static final int TP_NULL = 0;
    static final int TP_BOOL = 1;
    static final int TP_INT = 2;
    static final int TP_UINT = 3;
    static final int TP_FLOAT32 = 4;
    static final int TP_FLOAT64 = 5;
    static final int TP_STRING = 6;
    static final int TP_BLOB = 7;

    /**
     * Encode a TpValue into the bitstream.
     */
    public static void encodeValue(BitWriter writer, TpValue val) {
        if (val == null || val.getType() == TpValue.Type.NULL) {
            writer.writeBits(TP_NULL, 4);
            return;
        }

        switch (val.getType()) {
            case BOOL:
                writer.writeBits(TP_BOOL, 4);
                writer.writeBit(val.boolValue() ? 1 : 0);
                break;

            case INT:
                writer.writeBits(TP_INT, 4);
                VarInt.writeVarInt(writer, val.intValue());
                break;

            case UINT:
                writer.writeBits(TP_UINT, 4);
                VarInt.writeVarUint(writer, val.uintValue());
                break;

            case FLOAT64: {
                writer.writeBits(TP_FLOAT64, 4);
                long bits = Double.doubleToRawLongBits(val.float64Value());
                long hi = (bits >>> 32) & 0xFFFFFFFFL;
                long lo = bits & 0xFFFFFFFFL;
                writer.writeU64Parts(hi, lo);
                break;
            }

            case STRING: {
                byte[] encoded = val.stringValue().getBytes(StandardCharsets.UTF_8);
                writer.writeBits(TP_STRING, 4);
                VarInt.writeVarUint(writer, encoded.length);
                writer.alignToByte();
                writer.writeBytes(encoded);
                break;
            }

            case BLOB: {
                byte[] blob = val.blobValue();
                writer.writeBits(TP_BLOB, 4);
                VarInt.writeVarUint(writer, blob.length);
                writer.alignToByte();
                writer.writeBytes(blob);
                break;
            }

            default:
                throw new IllegalArgumentException("Unsupported value type: " + val.getType());
        }
    }

    /**
     * Decode a TpValue from the bitstream.
     */
    public static TpValue decodeValue(BitReader reader) {
        int tag = (int) reader.readBits(4);

        switch (tag) {
            case TP_NULL:
                return TpValue.ofNull();

            case TP_BOOL:
                return TpValue.ofBool(reader.readBit() != 0);

            case TP_INT:
                return TpValue.ofInt(VarInt.readVarInt(reader));

            case TP_UINT:
                return TpValue.ofUInt(VarInt.readVarUint(reader));

            case TP_FLOAT32: {
                long bits = reader.readU32();
                float f = Float.intBitsToFloat((int) bits);
                return TpValue.ofFloat64(f);
            }

            case TP_FLOAT64: {
                long hi = reader.readU32();
                long lo = reader.readU32();
                long bits = (hi << 32) | lo;
                return TpValue.ofFloat64(Double.longBitsToDouble(bits));
            }

            case TP_STRING: {
                int slen = (int) VarInt.readVarUint(reader);
                reader.alignToByte();
                byte[] raw = reader.readBytes(slen);
                return TpValue.ofString(new String(raw, StandardCharsets.UTF_8));
            }

            case TP_BLOB: {
                int blen = (int) VarInt.readVarUint(reader);
                reader.alignToByte();
                byte[] raw = reader.readBytes(blen);
                return TpValue.ofBlob(raw);
            }

            default:
                throw new IllegalArgumentException("Unknown value type tag: " + tag);
        }
    }
}

File: bindings/kotlin/src/main/kotlin/com/deftio/triepack/Values.kt
/*
 * Values.kt
 *
 * Value encoder/decoder for 8 type tags -- matches src/core/value.c.
 *
 * Copyright (c) 2026 M. A. Chatterjee
 * SPDX-License-Identifier: BSD-2-Clause
 */

package com.deftio.triepack

// Type tags (4-bit, matches tp_value_type enum)
const val TP_NULL: Int = 0
const val TP_BOOL: Int = 1
const val TP_INT: Int = 2
const val TP_UINT: Int = 3
const val TP_FLOAT32: Int = 4
const val TP_FLOAT64: Int = 5
const val TP_STRING: Int = 6
const val TP_BLOB: Int = 7

/**
 * Sealed class representing TriePack typed values.
 */
sealed class TpValue {
    object Null : TpValue() {
        override fun toString(): String = "TpValue.Null"
    }

    data class Bool(val value: Boolean) : TpValue()
    data class Int(val value: Long) : TpValue()
    data class UInt(val value: Long) : TpValue()
    data class Float64(val value: Double) : TpValue()
    data class Str(val value: String) : TpValue()
    data class Blob(val value: ByteArray) : TpValue() {
        override fun equals(other: Any?): Boolean {
            if (this === other) return true
            if (other !is Blob) return false
            return value.contentEquals(other.value)
        }

        override fun hashCode(): Int = value.contentHashCode()
    }
}

/**
 * Encode a TpValue into the bitstream.
 */
fun encodeValue(writer: BitWriter, value: TpValue?) {
    if (value == null || value is TpValue.Null) {
        writer.writeBits(TP_NULL.toLong(), 4)
        return
    }

    when (value) {
        is TpValue.Bool -> {
            writer.writeBits(TP_BOOL.toLong(), 4)
            writer.writeBit(if (value.value) 1 else 0)
        }
        is TpValue.Int -> {
            writer.writeBits(TP_INT.toLong(), 4)
            writeVarInt(writer, value.value)
        }
        is TpValue.UInt -> {
            writer.writeBits(TP_UINT.toLong(), 4)
            writeVarUint(writer, value.value)
        }
        is TpValue.Float64 -> {
            writer.writeBits(TP_FLOAT64.toLong(), 4)
            val bits = java.lang.Double.doubleToLongBits(value.value)
            val hi = (bits ushr 32) and 0xFFFFFFFFL
            val lo = bits and 0xFFFFFFFFL
            writer.writeU64(hi, lo)
        }
        is TpValue.Str -> {
            val encoded = value.value.toByteArray(Charsets.UTF_8)
            writer.writeBits(TP_STRING.toLong(), 4)
            writeVarUint(writer, encoded.size.toLong())
            writer.alignToByte()
            writer.writeBytes(encoded)
        }
        is TpValue.Blob -> {
            writer.writeBits(TP_BLOB.toLong(), 4)
            writeVarUint(writer, value.value.size.toLong())
            writer.alignToByte()
            writer.writeBytes(value.value)
        }
        is TpValue.Null -> { /* already handled above */ }
    }
}

/**
 * Decode a value from the bitstream.
 */
fun decodeValue(reader: BitReader): TpValue? {
    val tag = reader.readBits(4).toInt()

    return when (tag) {
        TP_NULL -> null
        TP_BOOL -> TpValue.Bool(reader.readBit() != 0)
        TP_INT -> TpValue.Int(readVarInt(reader))
        TP_UINT -> TpValue.UInt(readVarUint(reader))
        TP_FLOAT32 -> {
            val bits = reader.readU32()
            val floatVal = java.lang.Float.intBitsToFloat(bits.toInt())
            TpValue.Float64(floatVal.toDouble())
        }
        TP_FLOAT64 -> {
            val hi = reader.readU32()
            val lo = reader.readU32()
            val bits = (hi shl 32) or lo
            TpValue.Float64(java.lang.Double.longBitsToDouble(bits))
        }
        TP_STRING -> {
            val sLen = readVarUint(reader).toInt()
            reader.alignToByte()
            val raw = reader.readBytes(sLen)
            TpValue.Str(String(raw, Charsets.UTF_8))
        }
        TP_BLOB -> {
            val bLen = readVarUint(reader).toInt()
            reader.alignToByte()
            TpValue.Blob(reader.readBytes(bLen))
        }
        else -> throw IllegalArgumentException("Unknown value type tag: $tag")
    }
}
